File: features/xai-oauth/src/xai_oauth/grok_import.py
from __future__ import annotations

import json
import os
from datetime import UTC, datetime, timedelta
from pathlib import Path

from xai_oauth.constants import (
    CLIENT_ID,
    DEFAULT_EXPIRES_IN_SECS,
    GROK_SCOPE_KEY,
    ISSUER,
)
from xai_oauth.session import (
    AuthError,
    OAuthSession,
    claims_from_jwt,
    parse_expires_at,
)


def resolve_grok_auth_path(explicit: Path | None = None) -> Path:
    if explicit is not None:
        return explicit
    env_path = os.environ.get("GROK_AUTH_PATH")
    if env_path:
        return Path(env_path)
    grok_home = os.environ.get("GROK_HOME")
    if grok_home:
        return Path(grok_home) / "auth.json"
    return Path.home() / ".grok" / "auth.json"


def import_oauth_from_grok(path: Path | None = None) -> OAuthSession:
    auth_path = resolve_grok_auth_path(path)
    try:
        raw = auth_path.read_text(encoding="utf-8")
    except FileNotFoundError as e:
        raise AuthError(f"grok auth file not found: {auth_path}") from e
    except OSError as e:
        raise AuthError(f"cannot read grok auth file: {auth_path}") from e
    try:
        data = json.loads(raw)
    except json.JSONDecodeError as e:
        raise AuthError(f"corrupt grok auth file: {auth_path}") from e
    if not isinstance(data, dict):
        raise AuthError(f"corrupt grok auth file: {auth_path}")
    entry = data.get(GROK_SCOPE_KEY)
    if not isinstance(entry, dict):
        raise AuthError(f"no grok OAuth entry for {GROK_SCOPE_KEY}")
    access = entry.get("key")
    if not isinstance(access, str) or not access:
        raise AuthError("grok OAuth entry missing key")
    email = _optional_str(entry.get("email"))
    subject = _optional_str(entry.get("user_id")) or ""
    claims = claims_from_jwt(access)
    if not subject:
        sub = claims.get("sub")
        subject = sub if isinstance(sub, str) else ""
    if not subject:
        raise AuthError("grok OAuth entry missing user_id")
    if email is None:
        claim_email = claims.get("email")
        email = claim_email if isinstance(claim_email, str) else None
    expires_at = _expires_from_grok(entry, claims)
    return OAuthSession(
        access_token=access,
        refresh_token=None,
        expires_at=expires_at,
        subject=subject,
        email=email,
        issuer=ISSUER,
        client_id=CLIENT_ID,
    )


def _optional_str(value: object) -> str | None:
    if not isinstance(value, str) or not value:
        return None
    return value


def _expires_from_grok(entry: dict[str, object], claims: dict[str, object]) -> datetime:
    raw = entry.get("expires_at")
    if isinstance(raw, str) and raw:
        return parse_expires_at(raw)
    exp = claims.get("exp")
    if isinstance(exp, (int, float)) and not isinstance(exp, bool):
        return datetime.fromtimestamp(float(exp), tz=UTC)
    return datetime.now(UTC) + timedelta(seconds=DEFAULT_EXPIRES_IN_SECS)

File: crates/xai-oauth/src/grok_import.rs
use std::path::{Path, PathBuf};

use chrono::{Duration, TimeZone, Utc};
use serde_json::Value;

use crate::constants::{CLIENT_ID, DEFAULT_EXPIRES_IN_SECS, GROK_SCOPE_KEY, ISSUER};
use crate::error::AuthError;
use crate::session::{OAuthSession, claims_from_jwt};

pub(crate) fn resolve_grok_auth_path(explicit: Option<&Path>) -> PathBuf {
    if let Some(p) = explicit {
        return p.to_path_buf();
    }
    if let Ok(p) = std::env::var("GROK_AUTH_PATH") {
        return PathBuf::from(p);
    }
    if let Ok(home) = std::env::var("GROK_HOME") {
        return PathBuf::from(home).join("auth.json");
    }
    home_dir().join(".grok").join("auth.json")
}

pub(crate) fn import_oauth_from_grok(path: Option<&Path>) -> Result<OAuthSession, AuthError> {
    let auth_path = resolve_grok_auth_path(path);
    let raw = std::fs::read_to_string(&auth_path).map_err(|e| {
        if e.kind() == std::io::ErrorKind::NotFound {
            AuthError::msg(format!("grok auth file not found: {}", auth_path.display()))
        } else {
            AuthError::msg(format!(
                "cannot read grok auth file: {}",
                auth_path.display()
            ))
        }
    })?;
    let data: Value = serde_json::from_str(&raw)
        .map_err(|_| AuthError::msg(format!("corrupt grok auth file: {}", auth_path.display())))?;
    let obj = data.as_object().ok_or_else(|| {
        AuthError::msg(format!("corrupt grok auth file: {}", auth_path.display()))
    })?;
    let entry = obj
        .get(GROK_SCOPE_KEY)
        .and_then(Value::as_object)
        .ok_or_else(|| AuthError::msg(format!("no grok OAuth entry for {GROK_SCOPE_KEY}")))?;
    let access = entry
        .get("key")
        .and_then(Value::as_str)
        .filter(|s| !s.is_empty())
        .ok_or_else(|| AuthError::msg("grok OAuth entry missing key"))?
        .to_string();
    let mut email = entry
        .get("email")
        .and_then(Value::as_str)
        .filter(|s| !s.is_empty())
        .map(str::to_string);
    let mut subject = entry
        .get("user_id")
        .and_then(Value::as_str)
        .filter(|s| !s.is_empty())
        .unwrap_or("")
        .to_string();
    let claims = claims_from_jwt(&access);
    if subject.is_empty() {
        subject = claims
            .get("sub")
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string();
    }
    if subject.is_empty() {
        return Err(AuthError::msg("grok OAuth entry missing user_id"));
    }
    if email.is_none() {
        email = claims
            .get("email")
            .and_then(Value::as_str)
            .map(str::to_string);
    }
    let expires_at = expires_from_grok(entry, &claims)?;
    Ok(OAuthSession {
        access_token: access,
        refresh_token: None,
        expires_at,
        subject,
        email,
        issuer: ISSUER.to_string(),
        client_id: CLIENT_ID.to_string(),
    })
}

fn expires_from_grok(
    entry: &serde_json::Map<String, Value>,
    claims: &serde_json::Map<String, Value>,
) -> Result<chrono::DateTime<Utc>, AuthError> {
    if let Some(raw) = entry
        .get("expires_at")
        .and_then(Value::as_str)
        .filter(|s| !s.is_empty())
    {
        return chrono::DateTime::parse_from_rfc3339(raw)
            .map(|d| d.with_timezone(&Utc))
            .map_err(|e| AuthError::msg(format!("invalid expires_at: {e}")));
    }
    if let Some(exp) = claims.get("exp").and_then(Value::as_i64) {
        return Utc
            .timestamp_opt(exp, 0)
            .single()
            .ok_or_else(|| AuthError::msg("invalid exp claim"));
    }
    Ok(Utc::now() + Duration::seconds(DEFAULT_EXPIRES_IN_SECS))
}

fn home_dir() -> PathBuf {
    std::env::var_os("HOME")
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from("."))
}

#[cfg(test)]
#[path = "grok_import_tests.rs"]
mod tests;
